### PR TITLE
support loading upstream DNS servers when systemd-resolved stub resol…

### PR DESCRIPTION
**Pull request by knisbet**
…ver is enabled

Updates #606 
Requires Backport

The systemd-resolved stub resolver when configured points /etc/resolv.conf at 127.0.0.53 with a systemd-resolved listener. The configuration for upstream DNS appears to be embedded somewhere in DBUS, so it would require more research to see if there is a correct programatic way to pull the upstream DNS configuration into gravity. For now, it seems we can read a compatibility more resolv.conf from /run/systemd/resolve/resolv.conf which has the populated list of upstream DNS servers.

### CODE REVIEW

File: lib/install/phases/coredns.go [hunk 1]
```diff
@@ -93,12 +93,24 @@ func (r *corednsExecutor) Execute(ctx context.Context) error {
 	if err != nil {
 		return trace.Wrap(err)
 	}
+	servers := resolvConf.Servers
+
+	// Optionally try and load upstream nameservers from systemd-resolved by reading the compatibility resolv.conf
+	// More Info: https://github.com/gravitational/gravity/issues/606#issuecomment-529171440
+	// TODO(knisbet) is there a better way to pull upstream resolvers directly from systemd?
+	systemdResolvConf, err := systeminfo.ResolvFromFile("/run/systemd/resolve/resolv.conf")
```
Comment (a-palchikov) [inline lib/install/phases/coredns.go]: Even though I also have `systemd-resolved` active on my ubuntu `18.04`, I don't have any upstream nameservers listed in `/run/systemd/resolve/resolv.conf` - instead it just links back to `systemd-resolved` with `127.0.0.53`.
I'm not sure which of the [compatibility modes](https://www.freedesktop.org/software/systemd/man/systemd-resolved.service.html#/etc/resolv.conf) is on but the file does not list any link-specific DNS servers.

One way I saw this information being queried from `systemd-resolved` is by fetching `DNS` servers property for the specific link from [DBus](https://www.freedesktop.org/wiki/Software/systemd/resolved/). Here's an [example](https://github.com/FiloSottile/captive-browser/blob/master/cmd/systemd-networkd-dns/main.go) of the query.
Also, on my system the `resolve1` DBus service was not registered and I subsequently failed to query it with DBus although `systemd-resolve --status` is reporting the DNS servers properly.

Comment (knisbet) [reply lib/install/phases/coredns.go]: Well that sucks, I was hoping regardless of compatibility mode it was writing that file. thanks for checking. Guess we'll have to go the dbus route, but the problem I have with that is I found out how to pull the per interface DNS over dbug (although that go code is a good reference), I didn't see a clear indication on how to pull the global resolver when set...

```
Global
         DNS Servers: 8.8.8.8
          DNSSEC NTA: 10.in-addr.arpa
                      16.172.in-addr.arpa
                      168.192.in-addr.arpa
```

`systemd-resolved --status` obviously pulls it from somewhere, just need to waste time figuring out where that is.

**Comment by a-palchikov** [comment]
Ok, so after restarting the `NetworkManager` service on my system, the `systemd-resolved` did update the `/run/systemd/resolve/resolv.conf` file with the list of external nameservers and seems to keep them in sync if they're updated in the settings.

File: lib/install/phases/coredns.go (hunk 1 above)
Comment (a-palchikov) [reply lib/install/phases/coredns.go]: Ok, so after restarting the `NetworkManager` service on my system, the `systemd-resolved` did update the `/run/systemd/resolve/resolv.conf` file with the list of external nameservers and seems to keep them in sync if they're updated in the settings.

File: lib/install/phases/coredns.go [hunk 2]
```diff
@@ -93,12 +93,24 @@ func (r *corednsExecutor) Execute(ctx context.Context) error {
...
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+	if systemdResolvConf != nil {
+		servers = append(servers, systemdResolvConf.Servers...)
```
Comment (r0mant) [inline lib/install/phases/coredns.go]: Do we need to do some kind of deduplication on this list now?

**Comment by knisbet** [comment]
I request re-review, since I added deuplication logic. Also I think this change may need to be replicated into planet and it's resolv.conf as well.

**Pull request merged**